# File: 06_samtools_array_job.R
# Auth: umar.niazi@kcl.as.uk
# DESC: create a parameter file and shell script to run array job on hpc
# Date: 19/9/2019

## alternate file for processing original bam files from mcgill

## set variables and source libraries
source('header.R')

## connect to mysql database to get sample information
library('RMySQL')

##### connect to mysql database to get samples
db = dbConnect(MySQL(), user='rstudio', password='12345', dbname='Projects', host='127.0.0.1')
dbListTables(db)

# match the database id names with the bam files
g_did
q = paste0('select * from Sample where Sample.idData=41;')
dfQuery = dbGetQuery(db, q)
dfQuery$title = gsub(" ", "", dfQuery$title, fixed = T)
dfQuery

## extract original file names from description fields
cvDesc = as.character(dfQuery$description)
l = strsplit(cvDesc, ';')
cvDesc = lapply(l, function(x) return(x[2]))
cvDesc = unlist(cvDesc)
dfQuery$sn = cvDesc
rm(cvDesc)

## check if file names exist i.e. match 
csFiles = list.files('dataExternal/data/originalBams/', '*.bam', recursive = T)
identical(dfQuery$sn, csFiles)

# set header variables 
cvShell = '#!/bin/bash'
cvShell.2 = '#$ -S /bin/bash'
cvProcessors = '#$ -pe smp 1'
cvWorkingDir = '#$ -cwd'
cvJobName = '#$ -N samtools-array'
cvStdout = '#$ -j y'
cvMemoryReserve = '#$ -l h_vmem=2G'
cvArrayJob = paste0('#$ -t 1-', nrow(dfQuery))
# using low memory queue with one slot and 2 Gigs of memory

# set the directory names
cvInput = 'input/'
cvSam = '/opt/apps/bioinformatics/samtools/1.3.1/bin/samtools'

# create a parameter file and shell script
setwd('mcgillBams/')
dir.create('AutoScripts')
oFile.param = file('AutoScripts/samtools_param.txt', 'wt')

temp = lapply(seq_along(dfQuery$sn), function(x){
  # write samtools command variables
  fn = dfQuery$sn[x]
  in.s1 = paste0(cvInput, fn)
  # output file in bam format
  #s2 = paste0(cvInput, lf[[1]], '.bam')
  # remove low quality reads below 10
  s3 = paste0(cvInput, fn, '_q10.bam')
  # sort the file
  s4 = paste0(cvInput, fn, '_q10_sort.bam')
  # remove duplicates
  s5 = paste0(cvInput, fn, '_q10_sort_rd.bam')
  p1 = paste(in.s1, s3, s4, s5, sep=' ')
  writeLines(p1, oFile.param)
  return(data.frame(idSample=dfQuery$id[x], name=c(in.s1, s4, s5), type=c('mcgill original bam', 'mcgill quality 10 sorted bam',
                                                                             'mcgill quality 10 sorted bam duplicates removed'), 
                    group1=dfQuery$group1[x]))
})

close(oFile.param)
temp[sapply(temp, is.null)] = NULL
dfNewData = do.call(rbind, temp)
rownames(dfNewData) = NULL
# remove the word input/ from file name
dfNewData$name = gsub('input/(\\w+)', '\\1', as.character(dfNewData$name), perl=T)

oFile = file('AutoScripts/samtools.sh', 'wt')

writeLines(c('# Autogenerated script from write_samtools_script.R', paste('# date', date())), oFile)
writeLines(c('# make sure directory paths exist before running script'), oFile)
writeLines(c(cvShell, cvShell.2, cvProcessors, cvWorkingDir, cvJobName, cvStdout, cvMemoryReserve, cvArrayJob), oFile)
writeLines('\n\n', oFile)

# module load
writeLines(c('module load bioinformatics/samtools/1.3.1'), oFile)
writeLines('\n\n', oFile)

## write array job lines
writeLines("# Parse parameter file to get variables.
number=$SGE_TASK_ID
paramfile=samtools_param.txt
 
#ins1=`sed -n ${number}p $paramfile | awk '{print $1}'`
bamfile=`sed -n ${number}p $paramfile | awk '{print $1}'`
bamq10=`sed -n ${number}p $paramfile | awk '{print $2}'`
bamq10sort=`sed -n ${number}p $paramfile | awk '{print $3}'`
bamrd=`sed -n ${number}p $paramfile | awk '{print $4}'`

# 9. Run the program.", oFile)

# sam to bam
# p1 = paste('samtools view -b -S', '$ins1', '>', '$bamfile', sep=' ')
# com1 = paste(p1)
# remove low quality reads
p1 = paste('samtools view -b -q 10', '$bamfile', '>', '$bamq10', sep=' ')
com2 = paste(p1)
# sort the file
p1 = paste('samtools sort -o', '$bamq10sort', '$bamq10', sep=' ')
com3 = paste(p1)
# create index before duplicate removal
p1 = paste('samtools index', '$bamq10sort', sep=' ')
com3.1 = paste(p1)
# remove duplicates, for paired end reads
p1 = paste('samtools rmdup', '$bamq10sort', '$bamrd', sep=' ')
com4 = paste(p1)
# create index
p1 = paste('samtools index', '$bamrd', sep=' ')
com5 = paste(p1)

writeLines(c(com2, com3, com3.1, com4, com5), oFile)
writeLines('\n\n', oFile)

close(oFile)
dbDisconnect(db)

### update database with the file names
# db = dbConnect(MySQL(), user='rstudio', password='12345', dbname='Projects', host='127.0.0.1')
# dbWriteTable(db, name='File', value = dfNewData, append=T, row.names=F)
# dbDisconnect(db)
